internal/pkg/service/buffer/storage/local/reader/readchain/chain_test.go

```diff
@@ -0,0 +1,334 @@
+	t.Parallel()
+	tc := newChainTestCase(t)
+
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReader{inner: r, Logger: tc.Logger, Name: "R1"}
+	})
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReadCloser{inner: r, Logger: tc.Logger, Name: "RC2"}
+	})
+	tc.Chain.PrependCloseFn("FN1", func() error {
+		tc.Logger.Info(`TEST: close "FN1"`)
+		return nil
+	})
+	tc.Chain.PrependCloseFn("FN2", func() error {
+		tc.Logger.Info(`TEST: close "FN2"`)
+		return nil
+	})
+	tc.Chain.AppendCloseFn("FN3", func() error {
+		tc.Logger.Info(`TEST: close "FN3"`)
+		return nil
+	})
+	tc.Chain.AppendCloseFn("FN4", func() error {
+		tc.Logger.Info(`TEST: close "FN4"`)
+		return nil
+	})
+
+	// Read all from the Chain
+	content, err := io.ReadAll(tc.Chain)
+	if assert.NoError(tc.T, err) {
+		assert.Equal(tc.T, "foo bar", string(content))
+	}
+
+	// Close the chain
+	assert.NoError(t, tc.Chain.Close())
+
+	// 1st read is the content, 2nd is EOF error
+	tc.AssertLogs(`
+INFO  TEST: read "RC2"
+INFO  TEST: read "R1"
+INFO  TEST: read "file"
+INFO  TEST: read "RC2"
+INFO  TEST: read "R1"
+INFO  TEST: read "file"
+DEBUG  closing chain
+INFO  TEST: close "FN2"
+INFO  TEST: close "FN1"
+INFO  TEST: close "RC2"
+INFO  TEST: close "file"
+INFO  TEST: close "FN3"
+INFO  TEST: close "FN4"
+DEBUG  chain closed
+`)
+}
+
+// TestChain_ReadError tests a Read error.
+func TestChain_ReadError(t *testing.T) {
+	t.Parallel()
+	tc := newChainTestCase(t)
+
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReader{inner: r, Logger: tc.Logger, Name: "R1", ReadError: errors.New("some error")}
+	})
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReadCloser{inner: r, Logger: tc.Logger, Name: "RC2"}
+	})
+
+	// Read all from the Chain
+	_, err := io.ReadAll(tc.Chain)
+	if assert.Error(tc.T, err) {
+		assert.Equal(tc.T, "some error", err.Error())
+	}
+
+	tc.AssertLogs(`
+INFO  TEST: read "RC2"
+INFO  TEST: read "R1"
+`)
+}
+
+// TestChain_CloseError tests a Close error.
+func TestChain_CloseError(t *testing.T) {
+	t.Parallel()
+	tc := newChainTestCase(t)
+
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReadCloser{inner: r, Logger: tc.Logger, Name: "RC1"}
+	})
+	tc.Chain.PrependReader(func(r io.Reader) io.Reader {
+		return &testReadCloser{inner: r, Logger: tc.Logger, Name: "RC2", CloseError: errors.New("some error")}
+	})
+
+	// Read all from the Chain
+	content, err := io.ReadAll(tc.Chain)
+	if assert.NoError(tc.T, err) {
+		assert.Equal(tc.T, "foo bar", string(content))
+	}
+
+	// Read all from the Chain
+	err = tc.Chain.Close()
+	if assert.Error(tc.T, err) {
+		assert.Equal(tc.T, "chain close error: cannot close \"RC2\": some error", err.Error())
+	}
+
+	// 1st read is the content, 2nd is EOF error
+	tc.AssertLogs(`
+INFO  TEST: read "RC2"
+INFO  TEST: read "RC1"
+INFO  TEST: read "file"
+INFO  TEST: read "RC2"
+INFO  TEST: read "RC1"
+INFO  TEST: read "file"
```

Comment:
to sem asi nějak nepobral, proč sa volá ten read 2x ?

Reply:
`ReadAll` bezi kym nedostanes error EOF (end of file).
Teda prvy read vrati content (cely, kedze tam je kratky string, inak by sa read opakoval dookola), a posledny read vrati `n=0`, `err=EOF`.